File: server/datasets/java_source-code/Multiple-E4/type2_kmp_clone_1_substring_search.java
public class Solution {
    public int find_match(String main_str, String sub_str) {
        if (sub_str == null || sub_str.length() == 0) return 0;
        
        int[] prefix_arr = new int[sub_str.length()];
        int prev = 0, idx = 1;
        while (idx < sub_str.length()) {
            if (sub_str.charAt(idx) == sub_str.charAt(prev)) {
                prefix_arr[idx] = prev + 1;
                prev++;
                idx++;
            } else if (prev == 0) {
                prefix_arr[idx] = 0;
                idx++;
            } else {
                prev = prefix_arr[prev - 1];
            }
        }
        
        int h_idx = 0, n_idx = 0;
        while (h_idx < main_str.length()) {
            if (main_str.charAt(h_idx) == sub_str.charAt(n_idx)) {
                h_idx++;
                n_idx++;
            } else {
                if (n_idx == 0) {
                    h_idx++;
                } else {
                    n_idx = prefix_arr[n_idx - 1];
                }
            }
            
            if (n_idx == sub_str.length()) {
                return h_idx - sub_str.length();
            }
        }
        return -1;
    }
}

File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_e4/type2_kmp_clone_1_substring_search.py
def find_match(main_str, sub_str):
    if not sub_str: return 0
    
    prefix_arr = [0] * len(sub_str)
    prev, idx = 0, 1
    while idx < len(sub_str):
        if sub_str[idx] == sub_str[prev]:
            prefix_arr[idx] = prev + 1
            prev += 1
            idx += 1
        elif prev == 0:
            prefix_arr[idx] = 0
            idx += 1
        else:
            prev = prefix_arr[prev - 1]
            
    h_idx = n_idx = 0
    while h_idx < len(main_str):
        if main_str[h_idx] == sub_str[n_idx]:
            h_idx, n_idx = h_idx + 1, n_idx + 1
        else:
            if n_idx == 0: h_idx += 1
            else: n_idx = prefix_arr[n_idx - 1]
            
        if n_idx == len(sub_str):
            return h_idx - len(sub_str)
            
    return -1
